### PR TITLE
fix: detect 1.e or 1.E SQLI issue

### PR DESCRIPTION
**Pull request by kundu0914**
**ISSUES:**
https://github.com/libinjection/libinjection/issues/59

https://gosecure.ai/blog/2021/10/19/a-scientific-notation-bug-in-mysql-left-aws-waf-clients-vulnerable-to-sql-injection/ From this Bolg, we could see 1.e or 1.E is a risk SQLI, The SQL parser will ignore it during parsing.

Like "1.e(1)" => (1), 1 1.e/1 => 1/1 etc. see below a simple example 
image

So, if a payload such as "1' or 1.e(1)" bypass SQLI detection, which is really risk, we should detect such SQLI injection in case WAF bypass. 
The easiest solution maybe just ignore it instead of parsing it as TYPE_BAREWORD , the SQLI parser will continue to move to get the value after 1.e

**Analyze**
First, in some contexts, the 1.e definitely can be parsed as TYPE_BAREWORD, but it's no harm if we ignore it even though it should be parsed as TYPE_BAREWORD

Please think about the below cases:
E.g
input_data = token1 token2 1.2e token4 token5 token6 ....

Pasring xx.xxE as TYPE_BAREWORD
case 1: if input_data detects a SQLI, then it maybe a "false positive" detection if we parse 1.2e as TYPE_BAREWORD, since it could not be a SQLI after remove 1.2e. [we need fix]
case 2: if input_data not detects a SQLI, then it maybe a "false negative" detection if we parse 1.2e as TYPE_BAREWORD, since it could be a SQLI after remove 1.2e. [we need fix]

Ignoring xx.xxE
case 1: if input_data detects a SQLI, which is a "true positive" detection not false. Because after SQLI pass to SQL parser, xxx.xxE will be ignored, if we parsed it as TYPE_BAREWORD, then we may get a "false negative" detection [Fix the above case 2]
case 2: if input_data not detects a SQLI, which is a "true negative" detection. Why ? Let's assume it's a SQLI, after pass it to SQL parser, 1.2e will be ignored, and then execute the rest of SQL, but the rest of is not a valid SQL or SQLI. But if we parse 1.2e as TYPE_BAREWORD(n), then we may get a "false positive" detection, since xx.xxE should be ignored. [Fix the above case 1]

**Test**

1.Before fix 1.E issue, i run native libinjection tests, I saw the below results:

    kudu@sjc-mpfpl src % bash test-samples-sqli-negative.sh
    SQLI  : 21
    SAFE  : 400
    TOTAL : 421
    Threshold is 21, got 21, passing.

    kudu@sjc-mpfpl src % bash test-samples-sqli-positive.sh
    
    Threshold is 18, got 18, passing.
    kudu@sjc-mpfpl src % 

I believe you guys may need update these test samples files, positive/negative both reach the max failed counts

2.After fix 1.E issue, i run native libinjection tests again, it seems we get a little **better result for positive:)** 

    kudu@sjc-mpfpl src %  bash test-samples-sqli-negative.sh
    SQLI  : 21
    SAFE  : 400
    TOTAL : 421
    Threshold is 21, got 21, passing.

    kudu@sjc-mpfpl src %  bash test-samples-sqli-positive.sh
    Threshold is 18, got 17, passing.
    kudu@sjc-mpfpl src % 

3.Add a new sqli positive test file (sqli-1.e.txt) and added some negative cases for my fix, then rerun sqli tests, All tests passed.

    kudu@sjc-mpfpl src %  bash test-samples-sqli-negative.sh
    SQLI  : 21
    SAFE  : 402
    TOTAL : 423
    Threshold is 21, got 21, passing.

    kudu@sjc-mpfpl src %  bash test-samples-sqli-positive.sh
    Threshold is 18, got 17, passing.

### CODE REVIEW

**Comment by fzipi** [comment]
Thanks for your contribution.

Can you please add tests to this change?

**Comment by kundu0914** [comment]
@fzipi Sure, let me add some test under your test framework.

**Comment by kundu0914** [comment]
@fzipi  i didn't see any positive/negative test failures... 

Test

1.Before fix 1.E issue, i run native libinjection tests, I saw the below results:

kudu@sjc-mpfpl src % bash test-samples-sqli-negative.sh
SQLI  : 21
SAFE  : 400
TOTAL : 421
Threshold is 21, got 21, passing.

kudu@sjc-mpfpl src % bash test-samples-sqli-positive.sh

Threshold is 18, got 18, passing.
kudu@sjc-mpfpl src % 
I believe you guys may need update these test samples files, positive/negative both reach the max failed counts

2.After fix 1.E issue, i run native libinjection tests again, it seems we get a little better result for positive:)

kudu@sjc-mpfpl src %  bash test-samples-sqli-negative.sh
SQLI  : 21
SAFE  : 400
TOTAL : 421
Threshold is 21, got 21, passing.

kudu@sjc-mpfpl src %  bash test-samples-sqli-positive.sh
Threshold is 18, got 17, passing.
kudu@sjc-mpfpl src % 
3.Add a new sqli positive test file (sqli-1.e.txt) and added some negative cases for my fix, then rerun sqli tests, All tests passed.

kudu@sjc-mpfpl src %  bash test-samples-sqli-negative.sh
SQLI  : 21
SAFE  : 402
TOTAL : 423
Threshold is 21, got 21, passing.

kudu@sjc-mpfpl src %  bash test-samples-sqli-positive.sh
Threshold is 18, got 17, passing.

**Comment by fzipi** [comment]
I meant the pipeline isn't passing.

**Review by fzipi — APPROVED** [review]
Looks like now tests are passing.

@migueldemoura comments?

File: src/libinjection_sqli.c [hunk 1]
```diff
@@ -1172,16 +1172,21 @@ static size_t parse_number(struct libinjection_sqli_state *sf) {
...
+    /* very special form of
+     * "1234.e"
+     * "10.10E"
+     * ".E"
+     *
+     * https://gosecure.ai/blog/2021/10/19/a-scientific-notation-bug-in-mysql-left-aws-waf-clients-vulnerable-to-sql-injection/
+     * From this Bolg, we could see 1.e or 1.E is a risk SQLI, The SQL parser
+     * will ignore it during parsing. Like "1.e(1)" => (1), 1 1.e/1 => 1/1 etc.
+     * So, if a payload such as "1' or 1.e(1)" bypass SQLI detection, which is
+     * really risk, we should detect such SQLI injection in case WAF bypass.
```
Comment (airween) [inline src/libinjection_sqli.c]:
```suggestion
     * In this blog post, we can see that 1.e or 1.E is a risky SQLI. The SQL parser
     * ignores it during parsing. For example, "1.e(1)" => (1), 1 1.e/1 => 1/1, etc.
     * So, if a payload like "1' or 1.e(1)" bypasses SQLI detection, which is
     * really risky, then we should detect such SQLI injection in case of WAF bypass.
```
